File: tic_tac_cs/Board.cs
namespace tic_tac_cs
{
    interface IBoard
    {
        string[,] GetCells();
        bool IsValidPosition(int row, int col, bool fullCheck);
        string GetCell(int row, int col);
        void SetCell(int row, int col, string mark);
        void Display();
        bool IsFull();
        bool HasHorizontalWin();
        bool HasVerticalWin();
        bool HasDiagonalWin();
        bool CheckWinner();
        void SetRemainingCells();

    }

    internal class Board : IBoard
    {
        public const int SIZE = 3;
        private string[,] cells;

        public Board()
        {
            cells = new string[SIZE, SIZE];
            for (int i = 0; i < SIZE; i++)
            {
                for (int j = 0; j < SIZE; j++)
                {
                    cells[i, j] = string.Empty;
                }
            }
        }

        public string[,] GetCells()
        {
            return cells;
        }

        public bool CheckWinner()
        {
            return HasHorizontalWin() || HasVerticalWin() || HasDiagonalWin();
        }

        public void Display()
        {
            Console.WriteLine();
            Console.WriteLine("     c1  c2  c3");
            Console.WriteLine("    ------------");

            for (int i = 0; i < SIZE; i++)
            {
                Console.Write($"r{i + 1} |");
                for (int j = 0; j < SIZE; j++)
                {
                    Console.Write(string.Format(" {0,1} |", cells[i, j]));
                }

                Console.WriteLine();
                Console.WriteLine("   -------------");
            }

            Console.WriteLine();
        }

        public string GetCell(int row, int col)
        {
            if (!IsValidPosition(row, col, false))
            {
                throw new Exception("Invalid row or column value. Try again");
            }

            return cells[row, col];
        }

        public bool HasDiagonalWin()
        {
            List<string> mainDiagonal = new List<string>();
            List<string> secDiagonal = new List<string>();

            for (int i = 0; i < SIZE; i++)
            {
                mainDiagonal.Add(cells[i, i]);
                secDiagonal.Add(cells[i, SIZE - i - 1]);
            }

            bool mainMatch = !mainDiagonal.Contains("") && mainDiagonal.Distinct().Count() == 1;
            bool secMatch = !secDiagonal.Contains("") && secDiagonal.Distinct().Count() == 1;

            return mainMatch || secMatch;
        }

        public bool HasHorizontalWin()
        {
            for (int i = 0; i < SIZE; i++)
            {
                string[] row = Utils.GetRow(cells, i);

                if (!row.Contains("") && row.Distinct().Count() == 1)
                {
                    return true;
                }
            }

            return false;
        }

        public bool HasVerticalWin()
        {
            for (int i = 0; i < SIZE; i++)
            {
                string[] col = Utils.GetColumn(cells, i);

                if (!col.Contains("") && col.Distinct().Count() == 1)
                {
                    return true;
                }
            }
            return false;
        }

        public bool IsFull()
        {
            foreach (var cell in cells)
            {
                if (string.IsNullOrEmpty(cell))
                {
                    return false;
                }
            }

            return true;
        }

        public bool IsValidPosition(int row, int col, bool fullCheck = true)
        {
            if (row < 0 || col < 0 || row > SIZE || col > SIZE)
            {
                return false;
            }
            return !fullCheck || string.IsNullOrEmpty(cells[row, col]);
        }

        public void SetCell(int row, int col, string mark)
        {
            if (!IsValidPosition(row, col))
            {
                throw new Exception("Invalid row or column value. Try again");
            }
            else
            {
                cells[row, col] = mark;
            }
        }

        public void SetRemainingCells()
        {
            for (int i = 0; i < SIZE; i++)
            {
                for (int j = 0; j < SIZE; j++)
                {
                    if (string.IsNullOrEmpty(cells[i, j]))
                    {
                        SetCell(i, j, "-");
                    }
                }
            }
        }
    }
}

File: tic_tac_fs/Board.fs
module Board

open System

type Board() =
    let SIZE = 3
    member val cells = Array2D.create SIZE SIZE "" with get

    member this.IsValidPosition(row: int, col: int, ?fullCheck0: bool) : bool =
        let fullCheck = defaultArg fullCheck0 true
        
        if row < 0 || col < 0 || row > SIZE || col > SIZE then
            false
        else
            not fullCheck || String.IsNullOrEmpty(this.cells[row, col])

    member this.GetCell(row: int, col: int) : string =
        if not (this.IsValidPosition(row, col, false)) then
            raise (new Exception("Invalid row or column. Try again"))
        else
            this.cells[row, col]

    member this.SetCell(row: int, col: int, mark: string) = 
        if not (this.IsValidPosition(row, col)) then
            raise (new Exception("Invalid row or column. Try again"))
        else
            this.cells[row, col] <- mark

    member this.Display() =
        Console.WriteLine("\n     c1  c2  c3");
        Console.WriteLine("   -------------");

        for i in 0 .. SIZE - 1 do
            Console.Write($"r{i + 1} |");
            for j in 0 .. SIZE - 1 do
                Console.Write(String.Format(" {0,1} |", this.cells[i, j]));
            Console.WriteLine("\n   -------------")
        Console.WriteLine()

    member this.IsFull() : bool =
        let mutable isFull = true
        for i in 0 .. SIZE - 1 do
            for j in 0 .. SIZE - 1 do
                if String.IsNullOrEmpty(this.cells[i, j]) then
                    isFull <- false
        isFull

    member this.HasHorizontalWin() : bool =
        let mutable hasWin = false
        for i in 0 .. SIZE - 1 do
            let row = Utils.GetRow(this.cells, i)
            if not (Array.contains "" row) && Seq.distinct row |> Seq.length = 1 then
                hasWin <- true
        hasWin

    member this.HasVerticalWin() : bool =
        let mutable hasWin = false
        for i in 0 .. SIZE - 1 do
            let col = Utils.GetColumn(this.cells, i)
            if not (Array.contains "" col) && Seq.distinct col |> Seq.length = 1 then
                hasWin <- true
        hasWin

     member this.HasDiagonalWin() : bool =
        let mainDiagonal = [for i in 0 .. SIZE - 1 -> this.cells.[i, i]]
        let secDiagonal = [for i in 0 .. SIZE - 1 -> this.cells.[i, SIZE - i - 1]]

        let mainMatch = not (mainDiagonal |> List.contains "") && (mainDiagonal |> List.distinct |> List.length = 1)
        let secMatch = not (secDiagonal |> List.contains "") && (secDiagonal |> List.distinct |> List.length = 1)

        mainMatch || secMatch

     member this.CheckWinner() : bool =
        this.HasHorizontalWin() || this.HasVerticalWin() || this.HasDiagonalWin()

     member this.SetRemainingCells() =
        for i in 0 .. SIZE - 1 do
            for j in 0 .. SIZE - 1 do
                if String.IsNullOrEmpty(this.cells[i, j]) then
                    this.SetCell(i, j, Utils.DEF)
        
            
                    
